[JENKINS-63373] writeYaml should support writing a list as multiple d…

ian-twilightcoder opened the pull request:
…ocuments

Add a `datas` parameter to writeYaml as an alternate to `data`. `datas` must be a collection, and each element will be written as a YAML document in the file.

ian-twilightcoder commented:
(This time I have permission from my work to post this. Sorry about the mixup last time.)

rsandell commented on src/main/resources/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep/help.html:
```diff
@@ -34,7 +34,11 @@
-        A Mandatory Object containing the data to be serialized.
+        An Optional Object containing the data to be serialized.
+    </li>
+    <li>
```
It should be mentioned here that it's either `data` or `datas` that can be specified.

ian-twilightcoder replied:
Done

rsandell commented on src/main/resources/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep/config.jelly:
```diff
@@ -31,6 +31,9 @@
 	</f:entry>
+    <f:entry title="${%datas.title}" field="datas" description="${%datas.description}">
+        <f:textbox />
+    </f:entry>
```
Have you tested that this works? The datas parameter is likely databound to a string and won't convert to a `Collection` iirc. And also both `data` and `datas` will be specified in the form submission and will generate a code snippet that won't work.

I guess it is better to leave it out?

ian-twilightcoder replied:
I'm embarrassed to admit I didn't initially know what it was for. I see how it works now, and yeah I don't see a good way to make this work and it's not really necessary either. Removing!

rsandell commented on src/test/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStepTest.java:
```diff
@@ -140,7 +153,21 @@ public void writeNoData() throws Exception {
+         p.setDefinition(new CpsFlowDefinition(
+                 "node('slaves') {\n" +
+                         "  writeYaml file: 'test', data: ['a': 1, 'b': 2], datas: [['a': 1], ['b': 2]] \n" +
+                         "  def yml = readYaml file: 'test' \n" +
+                         "  assert yml == ['a','b','c'] \n"+
```
_nit:_ These lines aren't needed.

ian-twilightcoder replied:
Good point

rsandell commented on src/test/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStepUnitTest.java:
```diff
@@ -20,11 +21,26 @@ public void writeEmptyListObject() throws Exception {
     }
 
+    @Test @WithoutJenkins
+    public void writeEmptyListOfDocuments() throws Exception {
```
What is this testing? It doesn't look like it is testing anything at all.
Or do you mean to test that it doesn't throw an exception when you specify an empty list?

ian-twilightcoder replied:
It's meant to be an analogue of `writeEmptyListObject`, which is to test that it doesn't throw an exception on an empty list. Just this one passes the empty list as datas instead of data.

rsandell reviewed (changes requested):
Looks great over all, just some minor comments.

ian-twilightcoder commented:
I think I got all the comments addressed, is there anything else I need to do before this can be merged?

rsandell commented:
Sorry, I've been on vacation and trying to catch up after getting back.

rittneje commented:
@ian-twilightcoder Thanks for this change, would be useful for us as well. Can you make it work with the `returnText` flag I just added?

ian-twilightcoder commented:
> @ian-twilightcoder Thanks for this change, would be useful for us as well. Can you make it work with the `returnText` flag I just added?

Thanks for catching that, fixed up `ReturnTextExecution.run`. I'm not sure I got the constructor right though, I've had a tough time finding Stapler documentation for people who don't know anything about Stapler (me).

ian-twilightcoder commented:
> Sorry, I've been on vacation and trying to catch up after getting back.

@rsandell No problem. Got the merge conflicts fixed and I think this should be ready to go now.
